README.md
# bulma-taginputs
Bulma.io extension to add interaction on input tags


CHANGELOG.md
<a name="1.0.10"></a>
## [1.0.10](https://github.com/Wikiki/bulma-tagsinput/compare/1.0.9...1.0.10) (2018-05-31)


### Bug Fixes

* options on attach method ([942d936](https://github.com/Wikiki/bulma-tagsinput/commit/942d936))



<a name="1.0.9"></a>
## [1.0.9](https://github.com/Wikiki/bulma-tagsinput/compare/1.0.8...1.0.9) (2018-05-31)


### Bug Fixes

* issue [#49](https://github.com/Wikiki/bulma-tagsinput/issues/49) Potential Bug: TagsInput allows duplicate entries ([d88971a](https://github.com/Wikiki/bulma-tagsinput/commit/d88971a))



<a name="1.0.5"></a>
## [1.0.5](https://github.com/Wikiki/bulma-tagsinput/compare/1.0.4...1.0.5) (2018-02-21)



<a name="1.0.8"></a>
## [1.0.8](https://github.com/Wikiki/bulma-tagsinput/compare/1.0.7...1.0.8) (2018-05-12)


### Bug Fixes

* [#20](https://github.com/Wikiki/bulma-tagsinput/issues/20) Deleted tags result in leading commas ([25349a1](https://github.com/Wikiki/bulma-tagsinput/commit/25349a1))



<a name="1.0.7"></a>
## [1.0.7](https://github.com/Wikiki/bulma-tagsinput/compare/1.0.6...1.0.7) (2018-03-29)



<a name="1.0.6"></a>
## [1.0.6](https://github.com/Wikiki/bulma-tagsinput/compare/1.0.1...1.0.6) (2018-02-21)



<a name="1.0.5"></a>
## [1.0.5](https://github.com/Wikiki/bulma-tagsinput/compare/1.0.1...1.0.5) (2018-02-21)



<a name="1.0.4"></a>
## [1.0.4](https://github.com/Wikiki/bulma-tagsinput/compare/1.0.3...1.0.4) (2018-02-21)



<a name="1.0.3"></a>
## [1.0.3](https://github.com/Wikiki/bulma-tagsinput/compare/1.0.2...1.0.3) (2018-02-21)



<a name="1.0.2"></a>
## [1.0.2](https://github.com/Wikiki/bulma-tagsinput/compare/1.0.1...1.0.2) (2018-02-20)



<a name="1.0.1"></a>
## [1.0.1](https://github.com/Wikiki/bulma-tagsinput/compare/1.0.0...1.0.1) (2018-02-11)


### Bug Fixes

* **gulp:** Remove Bulma’s important from dist’sass ([5bb6c9c](https://github.com/Wikiki/bulma-tagsinput/commit/5bb6c9c))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/Wikiki/bulma-tagsinput/compare/0.2.7...1.0.0) (2018-02-10)


### Features

* **rollup:** Add Rollup transpilation ([548c633](https://github.com/Wikiki/bulma-tagsinput/commit/548c633))



<a name="0.2.7"></a>
## [0.2.7](https://github.com/Wikiki/bulma-tagsinput/compare/0.2.6...0.2.7) (2018-02-10)



<a name="0.2.6"></a>
## [0.2.6](https://github.com/Wikiki/bulma-tagsinput/compare/0.2.5...0.2.6) (2018-02-10)


### Bug Fixes

* **gulp:** Use UMD modules style ([3b5c8e6](https://github.com/Wikiki/bulma-tagsinput/commit/3b5c8e6))



<a name="0.2.5"></a>
## [0.2.5](https://github.com/Wikiki/bulma-tagsinput/compare/0.2.4...0.2.5) (2018-02-10)


### Bug Fixes

* **gulp:** Correct dist filenames ([60770e9](https://github.com/Wikiki/bulma-tagsinput/commit/60770e9))
* **package.json:** File paths ([71d2b72](https://github.com/Wikiki/bulma-tagsinput/commit/71d2b72))



<a name="0.2.4"></a>
## [0.2.4](https://github.com/Wikiki/bulma-tagsinput/compare/0.2.3...0.2.4) (2018-02-09)



<a name="0.2.3"></a>
## [0.2.3](https://github.com/Wikiki/bulma-tagsinput/compare/0.1.3...0.2.3) (2018-02-09)


### Bug Fixes

* **bower:** Add missing version ([cbe926e](https://github.com/Wikiki/bulma-tagsinput/commit/cbe926e))


### Features

* **gulp:** Add gulp dependencies ([c2d9f29](https://github.com/Wikiki/bulma-tagsinput/commit/c2d9f29))
* **gulp:** Add release task ([54eb049](https://github.com/Wikiki/bulma-tagsinput/commit/54eb049))



# Change Log

## [Unreleased](https://github.com/wikiki/bulma-tagsinput/tree/HEAD)

[Full Changelog](https://github.com/wikiki/bulma-tagsinput/compare/0.1.3...HEAD)

**Fixed bugs:**

- Tag selection by backspace key for empty input line not working \(Chrome\) [\#11](https://github.com/Wikiki/bulma-tagsinput/issues/11)
- Form field isn't accentuated if focused \(Bulma 0.6.1\) [\#10](https://github.com/Wikiki/bulma-tagsinput/issues/10)
- Selected tag isn't highlighted [\#8](https://github.com/Wikiki/bulma-tagsinput/issues/8)
- Tag selecting of key events is not working properly [\#7](https://github.com/Wikiki/bulma-tagsinput/issues/7)
- Selecting and deleting tags via keys not working in Chrome [\#6](https://github.com/Wikiki/bulma-tagsinput/issues/6)
- doesn't able to delete tag [\#4](https://github.com/Wikiki/bulma-tagsinput/issues/4)
- Can't create a tag with comma [\#3](https://github.com/Wikiki/bulma-tagsinput/issues/3)
- Can't delete tags with backspace key [\#2](https://github.com/Wikiki/bulma-tagsinput/issues/2)
- Update tagsinput.js [\#5](https://github.com/Wikiki/bulma-tagsinput/pull/5) ([nille](https://github.com/nille))

## [0.1.3](https://github.com/wikiki/bulma-tagsinput/tree/0.1.3) (2017-11-30)
**Merged pull requests:**

- Create LICENSE [\#1](https://github.com/Wikiki/bulma-tagsinput/pull/1) ([Wikiki](https://github.com/Wikiki))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

gulpfile.js
var package                     = require('./package.json')
var gulp                        = require('gulp');

// Define variables.
var autoprefixer                = require('autoprefixer');
var babel                       = require('gulp-babel');
var bump                        = require('gulp-bump');
var camelCase                   = require('camelcase');
var cleancss                    = require('gulp-clean-css');
var concat                      = require('gulp-concat');
var conventionalChangelog       = require('gulp-conventional-changelog');
var conventionalGithubReleaser  = require('conventional-github-releaser');
var del                         = require('del');
var fs                          = require('fs');
var git                         = require('gulp-git');
var gutil                       = require('gulp-util');
var postcss                     = require('gulp-postcss');
var rollup                      = require('gulp-better-rollup');
var runSequence                 = require('run-sequence');
var sass                        = require('gulp-sass');
var sourcemaps                  = require('gulp-sourcemaps');
var spawn                       = require('child_process').spawn;
var minify                      = require('gulp-babel-minify');

/**
 * ----------------------------------------
 *  VARIABLES
 * ----------------------------------------
 */
var paths = {
  src: 'src/',
  dest: 'dist/',
  bulma: 'node_modules/bulma/sass/utilities/',
  jsPattern: '**/*.js'
}
var bulmaSassFile  = '_all.sass';
var globalSassFile = package.name + '.sass';
var globalJsFile   = package.name + '.js';
var mainSassFile   = 'extension.sass';
var mainJsFile     = 'extension.js';
var distCssFile    = package.name + '.min.css';
var distJsFile     = package.name + '.min.js';

/**
 * ----------------------------------------
 *  STYLESHEETS
 * ----------------------------------------
 */

 // Uses Sass compiler to process styles, adds vendor prefixes, minifies, then
 // outputs file to the appropriate location.
 gulp.task('build:styles', ['build:styles:copy'], function() {
   return gulp.src([paths.bulma + bulmaSassFile, paths.src + mainSassFile])
     .pipe(concat(globalSassFile))
     .pipe(sass({
       style: 'compressed',
       includePaths: [paths.bulma]
     }))
     .pipe(concat(distCssFile))
     .pipe(postcss([autoprefixer({browsers: ['last 2 versions']})]))
     .pipe(cleancss())
     .pipe(gulp.dest(paths.dest));
 });

 // Copy original sass file to dist
 gulp.task('build:styles:copy', function() {
   return gulp.src(paths.src + mainSassFile)
     .pipe(concat(globalSassFile))
     .pipe(gulp.dest(paths.dest));
 });

gulp.task('clean:styles', function(callback) {
  del([
    paths.dest + mainSassFile,
    paths.dest + distCssFile
  ]);
  callback();
});

/**
 * ----------------------------------------
 *  JAVASCRIPT
 * ----------------------------------------
 */

// Concatenates and uglifies global JS files and outputs result to the
// appropriate location.
gulp.task('build:scripts', function() {
  return gulp
    .src([paths.src + paths.jsPattern])
    .pipe(sourcemaps.init({
      loadMaps: true
    }))
    .pipe(rollup({
        plugins: [babel({
          babelrc: false,
          sourceMaps: true,
          exclude: 'node_modules/**',
          presets: [
            ["@babel/preset-env",  {
              "modules": false,
              "targets": {
                "browsers": gutil.env.babelTarget ? gutil.env.babelTarget : ["last 2 versions"]
              }
            }]
          ]
        })]
      }, {
        format: gutil.env.jsFormat ? gutil.env.jsFormat : 'umd',
        name: camelCase(package.name)
      }
    ))
    .pipe(concat(globalJsFile))
    .pipe(gulp.dest(paths.dest))
    .pipe(concat(distJsFile))
    .pipe(minify().on('error', function(err) {
      gutil.log(gutil.colors.red('[Error]'), err.toString())
    }))
    .pipe(sourcemaps.write())
    .pipe(gulp.dest(paths.dest));
});

gulp.task('clean:scripts', function(callback) {
  del([
    paths.dest + mainJsFile,
    paths.dest + distJsFile
  ]);
  callback();
});

// Deletes the entire dist directory.
gulp.task('clean', ['clean:scripts', 'clean:styles'], function(callback) {
  del(paths.dest);
  callback();
});

/**
 * ----------------------------------------
 *  GLOBAL BUILD
 * ----------------------------------------
 */
gulp.task('build', function(callback) {
  runSequence('clean',
    ['build:scripts', 'build:styles'],
    callback);
});

/**
 * ----------------------------------------
 *  DEFAULT TASK
 * ----------------------------------------
 */
gulp.task('default', ['build']);

/**
 * ----------------------------------------
 *  GENERATE CHANGELOG
 * ----------------------------------------
 */
gulp.task('changelog', function () {
  return gulp.src('CHANGELOG.md')
    .pipe(conventionalChangelog({
      preset: 'angular'
    }))
    .pipe(gulp.dest('./'));
});

/**
 * ----------------------------------------
 *  GENERATE GITHUB RELEASE
 * ----------------------------------------
 */
gulp.task('github:release', function(done) {
  conventionalGithubReleaser({
    type: "oauth",
    token: process.env.GITHUB_TOKEN
  }, {
    preset: 'angular'
  }, done);
});

/**
 * ----------------------------------------
 *  UPDATE
 * ----------------------------------------
 */
gulp.task('bump-version', function () {
  return gulp.src(['./bower.json', './package.json'])
    .pipe(bump({type: gutil.env.type ? gutil.env.type : 'patch' })
    .on('error', gutil.log))
    .pipe(gulp.dest('./'));
});


gulp.task('github:commit', function () {
  return gulp.src('.')
    .pipe(git.add())
    .pipe(git.commit('[Prerelease] Bumped version number'));
});

gulp.task('github:push', function (cb) {
  git.push('origin', 'master', cb);
});

gulp.task('github:create-new-tag', function (cb) {
  var version = getPackageJsonVersion();
  git.tag(version, 'Created Tag for version: ' + version, function (error) {
    if (error) {
      return cb(error);
    }
    git.push('origin', 'master', {args: '--tags'}, cb);
  });

  function getPackageJsonVersion () {
    // We parse the json file instead of using require because require caches
    // multiple calls so the version number won't be updated
    return JSON.parse(fs.readFileSync('./package.json', 'utf8')).version;
  };
});

gulp.task('npm:publish', function (done) {
  spawn('npm', ['publish'], { stdio: 'inherit' }).on('close', done);
});

gulp.task('release', function (callback) {
  runSequence(
    'build',
    'bump-version',
    'changelog',
    'github:commit',
    'github:push',
    'github:create-new-tag',
    'github:release',
    'npm:publish',
    function (error) {
      if (error) {
        console.log(error.message);
      } else {
        console.log('RELEASE FINISHED SUCCESSFULLY');
      }
      callback(error);
    });
});


src/extension.js
const MOUSE_EVENTS = ['click', 'touchstart'];

const KEY_BACKSPACE = 8,
  KEY_TAB = 9,
  KEY_ENTER = 13,
  KEY_LEFT = 37,
  KEY_RIGHT = 39,
  KEY_DELETE = 46,
  KEY_COMMA = 188;

export default class Tagify {
  constructor(selector, options = {}) {
    let defaultOptions = {
      disabled: false,
      delimiter: ',',
      allowDelete: true,
      lowercase: false,
      uppercase: false,
      duplicates: true
    }

    this.element = typeof selector === 'string'
      ? document.querySelector(selector)
      : selector;
    // An invalid selector or non-DOM node has been provided.
    if (!this.element) {
      throw new Error('An invalid selector or non-DOM node has been provided.');
    }

    this.options = Object.assign({}, defaultOptions, options);
    if (this.element.dataset.hasOwnProperty('lowercase')) {
      this.options.lowercase = this.element.dataset('lowercase')
    }
    if (this.element.dataset.hasOwnProperty('uppercase')) {
      this.options.lowercase = this.element.dataset('uppercase')
    }
    if (this.element.dataset.hasOwnProperty('duplicates')) {
      this.options.lowercase = this.element.dataset('duplicates')
    }

    this.init();
  }

  /**
   * Initiate all DOM element containing tagsinput class
   * @method
   * @return {Array} Array of all TagsInput instances
   */
  static attach(selector = 'input[type="tags"]', options = {}) {
    let tagsinputsInstances = new Array();

    const tagsinputs = document.querySelectorAll(selector);
    [].forEach.call(tagsinputs, tagsinput => {
      setTimeout(() => {
        tagsinputsInstances.push(new Tagify(tagsinput, options));
      }, 100);
    });
    return tagsinputsInstances;
  }

  init() {
    if (!this.options.disabled) {
      this.tags = [];
      // The container will visually looks like an input
      this.container = document.createElement('div');
      this.container.className = 'tagsinput';
      this.container.classList.add('field');
      this.container.classList.add('is-grouped');
      this.container.classList.add('is-grouped-multiline');
      this.container.classList.add('input');

      let inputType = this.element.getAttribute('type');
    	if (!inputType || inputType === 'tags') {
    		inputType = 'text';
      }
      // Create an invisible input element so user will be able to enter value
      this.input = document.createElement('input');
      this.input.setAttribute('type', inputType);
      if (this.element.getAttribute('placeholder')) {
        this.input.setAttribute('placeholder', this.element.getAttribute('placeholder'));
      } else {
        this.input.setAttribute('placeholder', 'Add a Tag');
      }
      this.container.appendChild(this.input);

      let sib = this.element.nextSibling;
      this.element.parentNode[sib ? 'insertBefore':'appendChild'](this.container, sib);
      this.element.style.cssText = 'position:absolute;left:0;top:0;width:1px;height:1px;opacity:0.01;';
      this.element.tabIndex = -1;

      this.enable();
    }
  }

  enable() {
    if (!this.enabled && !this.options.disabled) {

      this.element.addEventListener('focus', () => {
        this.container.classList.add('is-focused');
        this.select((Array.prototype.slice.call(this.container.querySelectorAll('.tag:not(.is-delete)'))).pop());
      });

      this.input.addEventListener('focus', () => {
    		this.container.classList.add('is-focused');
    		this.select((Array.prototype.slice.call(this.container.querySelectorAll('.tag:not(.is-delete)'))).pop());
      });
      this.input.addEventListener('blur', () => {
    		this.container.classList.remove('is-focused');
    		this.select((Array.prototype.slice.call(this.container.querySelectorAll('.tag:not(.is-delete)'))).pop());
    		this.savePartial();
      });
      this.input.addEventListener('keydown', (e) => {
        let key = e.charCode || e.keyCode || e.which,
          selectedTag,
          activeTag = this.container.querySelector('.tag.is-active'),
          last = (Array.prototype.slice.call(this.container.querySelectorAll('.tag:not(.is-delete)'))).pop(),
          atStart = this.caretAtStart(this.input);

        if (activeTag) {
          selectedTag = this.container.querySelector('[data-tag="' + activeTag.innerHTML.trim() + '"]');
        }
        this.setInputWidth();

        if (key === KEY_ENTER || key === this.options.delimiter.charCodeAt(0) || key === KEY_COMMA || key === KEY_TAB) {
          if (!this.input.value && (key !== this.options.delimiter.charCodeAt(0) || key === KEY_COMMA)) {
            return;
          }
          this.savePartial();
        } else if (key === KEY_DELETE && selectedTag) {
    			if (selectedTag.nextSibling) {
            this.select(selectedTag.nextSibling.querySelector('.tag'));
          } else if (selectedTag.previousSibling) {
            this.select(selectedTag.previousSibling.querySelector('.tag'));
          }
    			this.container.removeChild(selectedTag);
          this.tags.splice(this.tags.indexOf(selectedTag.getAttribute('data-tag')), 1);
    			this.setInputWidth();
    			this.save();
        } else if (key === KEY_BACKSPACE) {
          if (selectedTag) {
            if (selectedTag.previousSibling) {
    				  this.select(selectedTag.previousSibling.querySelector('.tag'));
            } else if (selectedTag.nextSibling) {
    				  this.select(selectedTag.nextSibling.querySelector('.tag'));
            }
    				this.container.removeChild(selectedTag);
            this.tags.splice(this.tags.indexOf(selectedTag.getAttribute('data-tag')), 1);
    				this.setInputWidth();
    				this.save();
    			} else if (last && atStart) {
    				this.select(last);
    			} else {
    				return;
          }
        } else if (key === KEY_LEFT) {
    			if (selectedTag) {
    				if (selectedTag.previousSibling) {
    					this.select(selectedTag.previousSibling.querySelector('.tag'));
    				}
    			} else if (!atStart) {
    				return;
    			} else {
    				this.select(last);
    			}
    		}
    		else if (key === KEY_RIGHT) {
    			if (!selectedTag) {
            return;
          }
    			this.select(selectedTag.nextSibling.querySelector('.tag'));
    		}
    		else {
    			return this.select();
        }

        e.preventDefault();
        return false;
      });
      this.input.addEventListener('input', () => {
        this.element.value = this.getValue();
        this.element.dispatchEvent(new Event('input'));
      });
      this.input.addEventListener('paste', () => setTimeout(savePartial, 0));

      this.container.addEventListener('mousedown', (e) => { this.refocus(e); });
      this.container.addEventListener('touchstart', (e) => { this.refocus(e); });

      this.savePartial(this.element.value);

      this.enabled = true;
    }
  }

  disable() {
    if (this.enabled && !this.options.disabled) {
      this.reset();

      this.enabled = false;
    }
  }

  select(el) {
		let sel = this.container.querySelector('.is-active');
		if (sel) {
      sel.classList.remove('is-active');
    }
		if (el) {
      el.classList.add('is-active');
    }
  }

  addTag(text) {
    if (~text.indexOf(this.options.delimiter)) {
      text = text.split(this.options.delimiter);
    }
    if (Array.isArray(text)) {
      return text.forEach((text) => {
        this.addTag(text)
      });
    }

    let tag = text && text.trim();
    if (!tag) {
      return false;
    }

    if (this.options['lowercase'] == 'true') {
      tag = tag.toLowerCase();
    }
    if (this.options['uppercase'] == 'true') {
      tag = tag.toUpperCase();
    }
    if (this.options['duplicates'] || this.tags.indexOf(tag) === -1) {
      this.tags.push(tag);

      let newTagWrapper = document.createElement('div');
      newTagWrapper.className = 'control';
      newTagWrapper.setAttribute('data-tag', tag);

      let newTag = document.createElement('div');
      newTag.className = 'tags';
      newTag.classList.add('has-addons');

      let newTagContent = document.createElement('span');
      newTagContent.className = 'tag';
      newTagContent.classList.add('is-active');
      this.select(newTagContent);
      newTagContent.innerHTML = tag;

      newTag.appendChild(newTagContent);
      if (this.options.allowDelete) {
        let newTagDeleteButton = document.createElement('a');
        newTagDeleteButton.className = 'tag';
        newTagDeleteButton.classList.add('is-delete');
        MOUSE_EVENTS.forEach((event) => {
          newTagDeleteButton.addEventListener(event, (e) => {
          let selectedTag,
            activeTag = e.target.parentNode,
            last = (Array.prototype.slice.call(this.container.querySelectorAll('.tag'))).pop(),
            atStart = this.caretAtStart(this.input);

          if (activeTag) {
            selectedTag = this.container.querySelector('[data-tag="' + activeTag.innerText.trim() + '"]');
          }

          if (selectedTag) {
    				this.select(selectedTag.previousSibling);
    				this.container.removeChild(selectedTag);
            this.tags.splice(this.tags.indexOf(selectedTag.getAttribute('data-tag')), 1);
    				this.setInputWidth();
    				this.save();
    			}
    			else if (last && atStart) {
    				this.select(last);
    			}
    			else {
    				return;
          }
        });
      });
        newTag.appendChild(newTagDeleteButton);
      }
      newTagWrapper.appendChild(newTag);

      this.container.insertBefore(newTagWrapper, this.input);
    }
  }

  getValue() {
    return this.tags.join(this.options.delimiter);
  }

  setValue(value) {
    (Array.prototype.slice.call(this.container.querySelectorAll('.tag'))).forEach((tag) => {
      this.tags.splice(this.tags.indexOf(tag.innerHTML), 1);
      this.container.removeChild(tag);
    });
    this.savePartial(value);
  }

  setInputWidth() {
    let last = (Array.prototype.slice.call(this.container.querySelectorAll('.control'))).pop();

    if (!this.container.offsetWidth) {
      return;
    }
    this.input.style.width = Math.max(this.container.offsetWidth - (last ? (last.offsetLeft + last.offsetWidth) : 30) - 30, this.container.offsetWidth / 4) + 'px';
  }

  savePartial(value) {
    if (typeof value !== 'string' && !Array.isArray(value)) {
      value = this.input.value;
    }
    if (this.addTag(value) !== false) {
			this.input.value = '';
			this.save();
			this.setInputWidth();
    }
  }

  save() {
    this.element.value = this.tags.join(this.options.delimiter);
    this.element.dispatchEvent(new Event('change'));
  }

  caretAtStart(el) {
		try {
			return el.selectionStart === 0 && el.selectionEnd === 0;
		}
		catch(e) {
			return el.value === '';
		}
  }

  refocus(e) {
		if (e.target.classList.contains('tag')) {
      this.select(e.target);
    }
		if (e.target === this.input) {
      return this.select();
    }
		this.input.focus();
		e.preventDefault();
		return false;
  }

  reset() {
    this.tags = [];
  }

  destroy() {
    this.disable();
    this.reset();
    this.element = null;
  }
}


tagsinput.js
if (typeof Object.assign != 'function') {
  // Must be writable: true, enumerable: false, configurable: true
  Object.defineProperty(Object, "assign", {
    value: function assign(target, varArgs) { // .length of function is 2
      'use strict';
      if (target == null) { // TypeError if undefined or null
        throw new TypeError('Cannot convert undefined or null to object');
      }

      var to = Object(target);

      for (var index = 1; index < arguments.length; index++) {
        var nextSource = arguments[index];

        if (nextSource != null) { // Skip over if undefined or null
          for (var nextKey in nextSource) {
            // Avoid bugs when hasOwnProperty is shadowed
            if (Object.prototype.hasOwnProperty.call(nextSource, nextKey)) {
              to[nextKey] = nextSource[nextKey];
            }
          }
        }
      }
      return to;
    },
    writable: true,
    configurable: true
  });
}

const MOUSE_EVENTS = ['click', 'touchstart'];

const KEY_BACKSPACE = 8,
  KEY_TAB = 9,
  KEY_ENTER = 13,
  KEY_LEFT = 37,
  KEY_RIGHT = 39,
  KEY_DELETE = 46,
  KEY_COMMA = 188;

class Tagify {
  constructor(element, options = {}) {
    let defaultOptions = {
      disabled: false,
      delimiter: ',',
      allowDelete: true,
      lowercase: false,
      uppercase: false,
      duplicates: true
    }
    this.element = element;
    this.options = Object.assign({}, defaultOptions, options);

    this.init();
  }

  init() {
    if (!this.options.disabled) {
      this.tags = [];
      // The container will visually looks like an input
      this.container = document.createElement('div');
      this.container.className = 'tagsinput';
      this.container.classList.add('field');
      this.container.classList.add('is-grouped');
      this.container.classList.add('is-grouped-multiline');
      this.container.classList.add('input');

      let inputType = this.element.getAttribute('type');
    	if (!inputType || inputType === 'tags') {
    		inputType = 'text';
      }
      // Create an invisible input element so user will be able to enter value
      this.input = document.createElement('input');
      this.input.setAttribute('type', inputType);
      if (this.element.getAttribute('placeholder')) {
        this.input.setAttribute('placeholder', this.element.getAttribute('placeholder'));
      } else {
        this.input.setAttribute('placeholder', 'Add a Tag');
      }
      this.container.appendChild(this.input);

      let sib = this.element.nextSibling;
      this.element.parentNode[sib ? 'insertBefore':'appendChild'](this.container, sib);
      this.element.style.cssText = 'position:absolute;left:0;top:0;width:1px;height:1px;opacity:0.01;';
      this.element.tabIndex = -1;

      this.enable();
    }
  }

  enable() {
    if (!this.enabled && !this.options.disabled) {

      this.element.addEventListener('focus', () => {
        this.container.classList.add('is-focused');
        this.select((Array.prototype.slice.call(this.container.querySelectorAll('.tag:not(.is-delete)'))).pop());
      });

      this.input.addEventListener('focus', () => {
    		this.container.classList.add('is-focused');
    		this.select((Array.prototype.slice.call(this.container.querySelectorAll('.tag:not(.is-delete)'))).pop());
      });
      this.input.addEventListener('blur', () => {
    		this.container.classList.remove('is-focused');
    		this.select((Array.prototype.slice.call(this.container.querySelectorAll('.tag:not(.is-delete)'))).pop());
    		this.savePartial();
      });
      this.input.addEventListener('keydown', (e) => {
        let key = e.charCode || e.keyCode || e.which,
          selectedTag,
          activeTag = this.container.querySelector('.tag.is-active'),
          last = (Array.prototype.slice.call(this.container.querySelectorAll('.tag:not(.is-delete)'))).pop(),
          atStart = this.caretAtStart(this.input);

        if (activeTag) {
          selectedTag = this.container.querySelector('[data-tag="' + activeTag.innerHTML.trim() + '"]');
        }
        this.setInputWidth();

        if (key === KEY_ENTER || key === this.options.delimiter.charCodeAt(0) || key === KEY_COMMA || key === KEY_TAB) {
          if (!this.input.value && (key !== this.options.delimiter.charCodeAt(0) || key === KEY_COMMA)) {
            return;
          }
          this.savePartial();
        } else if (key === KEY_DELETE && selectedTag) {
    			if (selectedTag.nextSibling) {
            this.select(selectedTag.nextSibling.querySelector('.tag'));
          } else if (selectedTag.previousSibling) {
            this.select(selectedTag.previousSibling.querySelector('.tag'));
          }
    			this.container.removeChild(selectedTag);
          delete this.tags[this.tags.indexOf(selectedTag.getAttribute('data-tag'))];
    			this.setInputWidth();
    			this.save();
        } else if (key === KEY_BACKSPACE) {
          if (selectedTag) {
            if (selectedTag.previousSibling) {
    				  this.select(selectedTag.previousSibling.querySelector('.tag'));
            } else if (selectedTag.nextSibling) {
    				  this.select(selectedTag.nextSibling.querySelector('.tag'));
            }
    				this.container.removeChild(selectedTag);
            delete this.tags[this.tags.indexOf(selectedTag.getAttribute('data-tag'))];
    				this.setInputWidth();
    				this.save();
    			} else if (last && atStart) {
    				this.select(last);
    			} else {
    				return;
          }
        } else if (key === KEY_LEFT) {
    			if (selectedTag) {
    				if (selectedTag.previousSibling) {
    					this.select(selectedTag.previousSibling.querySelector('.tag'));
    				}
    			} else if (!atStart) {
    				return;
    			} else {
    				this.select(last);
    			}
    		}
    		else if (key === KEY_RIGHT) {
    			if (!selectedTag) {
            return;
          }
    			this.select(selectedTag.nextSibling.querySelector('.tag'));
    		}
    		else {
    			return this.select();
        }

        e.preventDefault();
        return false;
      });
      this.input.addEventListener('input', () => {
        this.element.value = this.getValue();
        this.element.dispatchEvent(new Event('input'));
      });
      this.input.addEventListener('paste', () => setTimeout(savePartial, 0));

      this.container.addEventListener('mousedown', (e) => { this.refocus(e); });
      this.container.addEventListener('touchstart', (e) => { this.refocus(e); });

      this.savePartial(this.element.value);

      this.enabled = true;
    }
  }

  disable() {
    if (this.enabled && !this.options.disabled) {
      this.reset();

      this.enabled = false;
    }
  }

  select(el) {
		let sel = this.container.querySelector('.is-active');
		if (sel) {
      sel.classList.remove('is-active');
    }
		if (el) {
      el.classList.add('is-active');
    }
  }

  addTag(text) {
    if (~text.indexOf(this.options.delimiter)) {
      text = text.split(this.options.delimiter);
    }
    if (Array.isArray(text)) {
      return text.forEach((text) => {
        this.addTag(text)
      });
    }

    let tag = text && text.trim();
    if (!tag) {
      return false;
    }

    if (this.element.getAttribute('lowercase') || this.options['lowercase'] == 'true') {
      tag = tag.toLowerCase();
    }
    if (this.element.getAttribute('uppercase') || this.options['uppercase'] == 'true') {
      tag = tag.toUpperCase();
    }
    if (this.element.getAttribute('duplicates') == 'true' || this.options['duplicates'] || this.tags.indexOf(tag) === -1) {
      this.tags.push(tag);

      let newTagWrapper = document.createElement('div');
      newTagWrapper.className = 'control';
      newTagWrapper.setAttribute('data-tag', tag);

      let newTag = document.createElement('div');
      newTag.className = 'tags';
      newTag.classList.add('has-addons');

      let newTagContent = document.createElement('span');
      newTagContent.className = 'tag';
      newTagContent.classList.add('is-active');
      this.select(newTagContent);
      newTagContent.innerHTML = tag;

      newTag.appendChild(newTagContent);
      if (this.options.allowDelete) {
        let newTagDeleteButton = document.createElement('a');
        newTagDeleteButton.className = 'tag';
        newTagDeleteButton.classList.add('is-delete');
        MOUSE_EVENTS.forEach((event) => {
          newTagDeleteButton.addEventListener(event, (e) => {
          let selectedTag,
            activeTag = e.target.parentNode,
            last = (Array.prototype.slice.call(this.container.querySelectorAll('.tag'))).pop(),
            atStart = this.caretAtStart(this.input);

          if (activeTag) {
            selectedTag = this.container.querySelector('[data-tag="' + activeTag.innerText.trim() + '"]');
          }

          if (selectedTag) {
    				this.select(selectedTag.previousSibling);
    				this.container.removeChild(selectedTag);
            delete this.tags[this.tags.indexOf(selectedTag.getAttribute('data-tag'))];
    				this.setInputWidth();
    				this.save();
    			}
    			else if (last && atStart) {
    				this.select(last);
    			}
    			else {
    				return;
          }
        });
      });
        newTag.appendChild(newTagDeleteButton);
      }
      newTagWrapper.appendChild(newTag);

      this.container.insertBefore(newTagWrapper, this.input);
    }
  }

  getValue() {
    return this.tags.join(this.options.delimiter);
  }

  setValue(value) {
    (Array.prototype.slice.call(this.container.querySelectorAll('.tag'))).forEach((tag) => {
      delete this.tags[this.tags.indexOf(tag.innerHTML)];
      this.container.removeChild(tag);
    });
    this.savePartial(value);
  }

  setInputWidth() {
    let last = (Array.prototype.slice.call(this.container.querySelectorAll('.control'))).pop();

    if (!this.container.offsetWidth) {
      return;
    }
    this.input.style.width = Math.max(this.container.offsetWidth - (last ? (last.offsetLeft + last.offsetWidth) : 30) - 30, this.container.offsetWidth / 4) + 'px';
  }

  savePartial(value) {
    if (typeof value !== 'string' && !Array.isArray(value)) {
      value = this.input.value;
    }
    if (this.addTag(value) !== false) {
			this.input.value = '';
			this.save();
			this.setInputWidth();
    }
  }

  save() {
    this.element.value = this.tags.join(this.options.delimiter);
    this.element.dispatchEvent(new Event('change'));
  }

  caretAtStart(el) {
		try {
			return el.selectionStart === 0 && el.selectionEnd === 0;
		}
		catch(e) {
			return el.value === '';
		}
  }

  refocus(e) {
		if (e.target.classList.contains('tag')) {
      this.select(e.target);
    }
		if (e.target === this.input) {
      return this.select();
    }
		this.input.focus();
		e.preventDefault();
		return false;
  }

  reset() {
    this.tags = [];
  }

  destroy() {
    this.disable();
    this.reset();
    this.element = null;
  }
}

document.addEventListener( 'DOMContentLoaded', function () {
  let tagInputs = document.querySelectorAll('input[type="tags"]');
  [].forEach.call(tagInputs, function(tagInput) {
      new Tagify(tagInput);
  });
});
